cpp/src/arrow/util/iterator.h

```diff
@@ -35,6 +36,13 @@
 
 namespace arrow {
 
+namespace detail {
+
+template <typename Signature>
+using result_of_t = typename std::result_of<Signature>::type;
```

Comment:
It was unused in iterator.h so I just removed it.  I think I pulled it in because I was using Loop as a base for the transforming iterator but it mutated enough to longer require this.